File: src/main/java/io/github/mg138/tsbook/listener/event/damage/DamageEventHandler.java
package io.github.mg138.tsbook.listener.event.damage;

import io.github.mg138.tsbook.Book;
import io.github.mg138.tsbook.entities.effect.data.StatusEffect;
import io.github.mg138.tsbook.entities.effect.data.StatusEffectType;
import io.github.mg138.tsbook.items.ItemStats;
import io.github.mg138.tsbook.items.data.stat.StatType;
import io.github.mg138.tsbook.entities.effect.EffectHandler;

import io.github.mg138.tsbook.items.data.stat.set.DefenseType;
import io.github.mg138.tsbook.players.ArcticGlobalDataService;
import io.github.mg138.tsbook.players.data.PlayerData;
import io.lumine.xikage.mythicmobs.MythicMobs;
import io.lumine.xikage.mythicmobs.api.bukkit.BukkitAPIHelper;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.event.player.PlayerQuitEvent;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DamageEventHandler implements Listener {
    private static final BukkitAPIHelper mythicMobHelper = MythicMobs.inst().getAPIHelper();
    private static final Map<Entity, Long> lastDamageTime = new HashMap<>();

    public static void unload() {
        lastDamageTime.clear();
    }

    @EventHandler
    public void onEntityDeath(EntityDeathEvent event) {
        lastDamageTime.remove(event.getEntity());
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        DamageHandler.damageCD.remove(event.getPlayer());
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onDamage(EntityDamageEvent event) {
        Entity entity = event.getEntity();
        if (!(entity instanceof LivingEntity)) return;
        if (entity instanceof ArmorStand || entity instanceof Item) return;

        LivingEntity livingEntity = (LivingEntity) entity;

        livingEntity.setMaximumNoDamageTicks(0);
        livingEntity.setNoDamageTicks(0);

        switch (event.getCause()) {
            case FIRE_TICK:
            case FIRE:
            case LAVA:
                event.setCancelled(true);

                if ((System.currentTimeMillis() - lastDamageTime.getOrDefault(entity, 0L)) > 500) {
                    DamageHandler.simpleDamage(livingEntity, 2);
                    lastDamageTime.put(entity, System.currentTimeMillis());
                }
                return;
            case FALL:
                if (EffectHandler.hasEffect(livingEntity, StatusEffectType.FALL_DAMAGE_RESISTANCE)) {
                    event.setCancelled(true);

                    StatusEffect effect = EffectHandler.getEffect(livingEntity, StatusEffectType.FALL_DAMAGE_RESISTANCE);
                    assert effect != null;
                    DamageHandler.simpleDamage(livingEntity, event.getDamage() * (1 - effect.power));
                    return;
                }
        }

        if (event instanceof EntityDamageByEntityEvent) {
            if (mythicMobHelper.isMythicMob(livingEntity)) {
                ConfigurationSection mob = Book.Companion.getCfg().getMMMob(mythicMobHelper.getMythicMobInstance(entity).getType().getInternalName());
                if (mob != null) {
                    Map<StatType, Double> map = new HashMap<>();

                    for (String literalType : mob.getKeys(false)) {
                        StatType type = StatType.valueOf(literalType.toUpperCase());
                        if (DefenseType.DEFENSE.contains(type)) map.put(type, mob.getDouble(literalType));
                    }

                    DamageHandler.damagedByEntity((EntityDamageByEntityEvent) event, map);
                    return;
                }
            } else if (livingEntity instanceof Player) {
                Player player = (Player) livingEntity;
                List<ItemStats> stats = new ArrayList<>();
                PlayerData data = ArcticGlobalDataService.dataServiceInstance.getData(player, ArcticGlobalDataService.Companion.getPlayerDataRef());
                if (data != null) {
                    data.getEquipment().forEach((i, armor) -> stats.add(armor.getStats()));
                }
                DamageHandler.damagedByEntity((EntityDamageByEntityEvent) event, DamageHandler.getDefense(stats.toArray(new ItemStats[0])));
                return;
            }
            DamageHandler.damagedByEntity((EntityDamageByEntityEvent) event, new HashMap<>());
        }
    }
}

File: src/main/kotlin/io/github/mg138/tsbook/listener/event/damage/DamageEventHandler.kt
package io.github.mg138.tsbook.listener.event.damage

import io.github.mg138.tsbook.entity.effect.EffectHandler
import io.github.mg138.tsbook.listener.event.damage.DamageHandler.simpleDamage
import io.github.mg138.tsbook.listener.event.damage.DamageHandler.damagedByEntity
import org.bukkit.event.entity.EntityDeathEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.event.EventPriority
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.EntityDamageEvent.DamageCause
import io.github.mg138.tsbook.entity.effect.data.StatusType
import org.bukkit.event.entity.EntityDamageByEntityEvent
import io.github.mg138.tsbook.attribute.stat.StatMap
import io.github.mg138.tsbook.attribute.stat.util.StatUtil
import io.github.mg138.tsbook.players.ArcticGlobalDataService
import io.github.mg138.tsbook.players.data.PlayerData
import io.github.mg138.tsbook.setting.mob.MobConfig
import org.bukkit.entity.*
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import java.util.*
import kotlin.math.max

object DamageEventHandler : Listener {
    private val damageCD: MutableMap<Entity, Long> = HashMap()

    fun unload() {
        damageCD.clear()
        DamageHandler.unload()
    }

    @EventHandler
    fun onEntityDeath(event: EntityDeathEvent) {
        damageCD.remove(event.entity)
    }

    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        val player = event.player
        DamageHandler.remove(player)
        damageCD.remove(player)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onDamage(event: EntityDamageEvent) {
        val entity = event.entity
        if (entity !is LivingEntity || entity is ArmorStand || entity is Item) return

        when (event.cause) {
            DamageCause.FIRE_TICK, DamageCause.FIRE, DamageCause.LAVA -> {
                event.isCancelled = true
                if (System.currentTimeMillis() - damageCD.getOrDefault(entity, 0L) > 500) {
                    simpleDamage(entity, 2.0)
                    damageCD[entity] = System.currentTimeMillis()
                }
                return
            }
            DamageCause.FALL -> {
                EffectHandler.getEffect(entity, StatusType.FALL_DAMAGE_RESISTANCE)?.let {
                    event.isCancelled = true
                    simpleDamage(
                        entity,
                        StatUtil.calculateModifier(
                            event.damage,
                            max((1 - it.entityStatus.status.power), 0.0)
                        )
                    )
                    return
                }
            }
            else -> Unit
        }
        if (event is EntityDamageByEntityEvent) {
            when {
                DamageHandler.mythicMobHelper.isMythicMob(entity) -> {
                    MobConfig[DamageHandler.mythicMobHelper.getMythicMobInstance(entity).type.internalName]?.let {
                        damagedByEntity(event, it.stats)
                        return
                    }
                }
                entity is Player -> {
                    val stats = StatMap()
                    ArcticGlobalDataService.inst.getData<PlayerData>(entity, PlayerData::class)
                        ?.equipment?.forEach { _, armor ->
                            armor.itemStat?.let {
                                stats.putAll(StatUtil.getDefense(it))
                            }
                        }
                    damagedByEntity(event, stats)
                    return
                }
            }
            damagedByEntity(event, StatMap())
        }
    }
}